libraries/Wire/src/utility/twi.c

```diff
@@ -373,7 +394,14 @@ void twi_stop(void)
 
   // wait for stop condition to be exectued on bus
   // TWINT is not set after a stop condition!
+  uint32_t counter = 0;
   while(TWCR & _BV(TWSTO)){
+    counter++;
+    if((twi_timeout_us > 0ul) && (counter >= 25000)) {
```

Comment:
Why does this use a counter and not the micros-based timeout like elsewhere?

Reply:
Hmm. I assume that's leftover from when the dealy only had millisecond resoution and was needed to get a sort delay on the order of a millisecond (but it's not obvious as to why @wmarkow wanted a fixed and counter based timeout here. The only reason I can think of is if maybe there's something I don't know about `micros()` not working here). Now that I've rewritten it using `micros()`, you're right, it should be changed to match. I'll do that.

Reply:
Change merged in 68fe5f1

Reply:
Guys, please hold on. It uses counter in purpose. **twi_stop** method is being executed from **ISR(TWI_vect)**, which is an **interrupt service routine**. In my opinion - **millis** and **micros** will not work when executed from ISR, because they are updated by some othe timer ISR. Please get familiar with this conversations:
https://github.com/wmarkow/Arduino/commit/afc25a64316e38f29cf887f546b8a09c73375838#comments

Reply:
Thank you for the heads up.

Before making this change I actually looked at the millis() documentation explicitly to check for this exact limitation. It's not documented there in the notes save warnings section https://www.arduino.cc/reference/en/language/functions/time/millis/ 

I will change this back to a volatile counter and test the whole PR with my hardware tomorrow (luckily I decided to take it home with me during the pandemic lockdown).

Reply:
>looked at the millis() documentation explicitly to check for this exact limitation. It's not documented there

It is documented in the `attachInterrupt()` reference page:
https://www.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/#_about_interrupt_service_routines
>`millis()` relies on interrupts to count, so it will never increment inside an ISR. Since `delay()` requires interrupts to work, it will not work if called inside an ISR. `micros()` works initially but will start behaving erratically after 1-2 ms. `delayMicroseconds()` does not use any counter, so it will work as normal.

Of course, that might not be the obvious place to look for this information when you're not using `attachInterrupt()`.

Reply:
Good point about the ISR, so we cannot use `micros()` there. A hardcoded loop counter would work, but then the timeout is not longer configurable. We could calculate the loop count from the timeout value, but with the original implementation the end result really depends on what assembly the compiler produces (which can be unpredictable, see e.g. #118 for the same problem in `delayMicroseconds()`).

A better fix could be to handcode the entire waiting loop in assembly, then you know exactly how many cycles each loop uses and can calculate the loop count from the timeout and `F_CPU`. For a similar approach, see again #118 which has some links to fixed `delayMicroseconds()` implementations that fix this.

A less accurate alternative could be to put a delay in the loop, e.g. `delayMicroseconds(10)` or something and then loop for `timeout / 10` times. Loop overhead is probably less than a microsecond, so the actual timeout will not be perfect but close (if we also fix #118 to make `delayMicroseconds()` accurate again).

If `twi_stop()` is also executed outside of the ISR, then the above solutions will lead to a longer timeout when other interrupts occur (since then time effectively stops, unlike when you use `micros()` that uses a timer that keeps running during ISRs). Might not be a big issue, though.

We could also consider lowering the timeout here, since we are called from an ISR which should not really block execution for too long (`micros()` should work in ISRs for at least 1024us). We might need to see in what conditions this timeout could occur and if it is ever expected to genuinly take long?

One more alternative is to actually re-enable interrupts in the ISR (and disable the TWI interrupt to prevent calling ourselves again). This is something that might be fragile, but I think it might work properly when we're careful. This approach would allow other ISRs to run (such as the timer overflow ISR, but also other ISRs that might be time critical are no longer blocked by the timeout).

One last alternative, would be to actually fix `micros()` (and probably `millis()`) to work properly even when called from an ISR. This is probably not very difficult by just manually calling the ISR when the overflow flag is set but interrupts are disabled (we did a similar thing to HardwareSerial). See #232 for discussion about this.

Reply:
>A less accurate alternative could be to put a delay in the loop, e.g. delayMicroseconds(10) or something and then loop for timeout / 10 times. Loop overhead is probably less than a microsecond, so the actual timeout will not be perfect but close (if we also fix #118 to make delayMicroseconds() accurate again).

This is nice idea (when #232 are not fixed yet), even if its accuracy is not perfect (because of #118). To fix the freeze it is better to have some timeout than nothing. For me - for example - waiting 1000us +/- 100us is acceptable. The main point is that the TWI freeze doesn't lock the whole device. At the moment I can not point the example where the solution must be very accurate. Of course the timeout should not be too small, so the mechanism will not report false timeouts. I'm not so familiar with with TWI specification but in my tests I have set 10ms and my device was able to recover from the frozen state.

>We could also consider lowering the timeout here, since we are called from an ISR which should not really block execution for too long (micros() should work in ISRs for at least 1024us). We might need to see in what conditions this timeout could occur and if it is ever expected to genuinly take long?

The question is what is the minimal acceptable timeout value here, so the library will not report false timouts whene eveything is going well. When all TWI devices behave nice then the timeout will not occur. When something goes wrong, then the timeout will be raised and some additional actions may be taken: like resetting TWI master, resetting TWI slaves, etc.

Reply:
68fe5f1 is reverted via a9929dc. So we're back to using a hard coded, counter-based timeout in `twi_stop()` because it's called from an ISR where the `micros()` function does not work.

Reply:
And 849c97c makes the counter volatile so the compiler won't optimize it away.

Reply:
My favorite approach you list @matthijskooijman is to make `micros()` work in ISRs. At which point we can just go back to 68fe5f1.

On a related note, I've just had a quick review about what's actually being called by the ISR here and I see
`twi_onSlaveReceive(twi_rxBuffer, twi_rxBufferIndex);`
and
` twi_onSlaveTransmit();`

Those are arbitrary user code that gets run right in the ISR, right?

Reply:
Also see #328, which actually suggests re-enabling interrupts in `twi_stop`, so maybe that is a viable approach, then.

Reply:
So if #328 is fixed in the way, that it enables the interrupts, then **millis** and **micros** should work correctly in **twi_stop**? According to the Atmega328p datasheet, timer interrupts have higher priority than TWI, so it should work. Am I right? Or is there some catch here? :)

Reply:
Nothing's really progressing here, so I guess I'll assume that interrupts won't work in the ISR and calls from it. I'll go ahead and start implementing something that takes the user supplied timeout and figures out how many calls to `delayMicroseconds(10)` should be made to approximate the timeout the user specifies here.

Reply:
Sounds like a plan. I would suggest using `_delay_us()` instead of `delayMicroseconds()`, though, since the latter is still unreliable (I did some more work on that a few weeks ago, but haven't prepared a PR yet). `_delay_us()` is from avr-libc and only works on compile-time constants, but gives cycle-perfect delays.

Reply:
as of da7d85a, I now call `_delay_us(10)` as many times as needed to approximate the user's specified timeout here